### PR TITLE
FIX: NiBabel 5, and NetworkX 3 and DIPY 1.6 compatibility

### LINKED ISSUE #3530: networkx >= 3.0 unsupported
to_scipy_sparse_matrix will be deprecated, use to_scipy_sparse_array instead

As stated in networkx:
```
warnings.warn(
        (
            "\n\nThe scipy.sparse array containers will be used instead of matrices\n"
            "in Networkx 3.0. Use `to_scipy_sparse_array` instead."
        ),
        DeprecationWarning,
        stacklevel=2,
    )
```

Either nipype code is adapted to handle network 2 and 3 or setup.py should specify that networkx must be < 3 in the dependencies.

### LINKED ISSUE #3537: Build failure with nibabel 5.0
### Summary
The Debian package of nipype has errors in build time test since nibabel 5.0 was uploaded.

### Actual behavior
There is a [build log](https://salsa.debian.org/med-team/nipype/-/jobs/3861969) showing the errors (including all used preconditions and versions).  Here is the short test summary info:

    =========================== short test summary info ============================
    FAILED ../nipype/algorithms/tests/test_ErrorMap.py::test_errormap - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    FAILED ../nipype/algorithms/tests/test_metrics.py::test_fuzzy_overlap - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    ERROR ../nipype/algorithms/tests/test_TSNR.py::TestTSNR::test_tsnr - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    ERROR ../nipype/algorithms/tests/test_TSNR.py::TestTSNR::test_tsnr_withpoly1 - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    ERROR ../nipype/algorithms/tests/test_TSNR.py::TestTSNR::test_tsnr_withpoly2 - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    ERROR ../nipype/algorithms/tests/test_TSNR.py::TestTSNR::test_tsnr_withpoly3 - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    ERROR ../nipype/algorithms/tests/test_TSNR.py::TestTSNR::test_warning - ValueError: Image data has type int64, which may cause incompatibilities with other tools. To use this type, pass an explicit header or dtype argument to Nifti1Image().
    = 2 failed, 2718 passed, 243 skipped, 1 deselected, 5 xfailed, 229 warnings, 5 errors in 163.84s (0:02:43) =

### Expected behavior
Passing all tests also with nibabel 5.0

### How to replicate the behavior
Please verify the latest version of nibabel to build nipype.

### Platform details:

OS: Debian sid
You can find all information about Python version and nipype version in the [full build log](https://salsa.debian.org/med-team/nipype/-/jobs/3861969)

Kind regards, Andreas.

### LINKED ISSUE #3539: 18 test failures (`IndexError` and `AssertionError`) on `1.8.4`
Following [tentative resolution of networkx update](https://github.com/nipy/nipype/pull/3535), 18 tests still fail.
Full build and test log: https://ppb.chymera.eu/8412f1.log

### Platform details:

```
{'commit_hash': '%h',
 'commit_source': 'archive substitution',
 'networkx_version': '3.0',
 'nibabel_version': '3.2.1',
 'nipype_version': '1.8.4',
 'numpy_version': '1.24.0',
 'pkg_path': '/usr/lib/python3.10/site-packages/nipype',
 'scipy_version': '1.10.0',
 'sys_executable': '/usr/lib/python-exec/python3.10/python',
 'sys_platform': 'linux',
 'sys_version': '3.10.9 (main, Dec 18 2022, 05:30:37) [GCC 12.2.1 20221210]',
 'traits_version': '6.3.2'}
```

### Execution environment

- My python environment outside container

### PR DESCRIPTION
**Pull request by effigies**
## Summary

This PR makes minimal changes to handle deprecations in nibabel 5 and networkx 3.

As seen in #3535, simply switching to sparse arrays breaks the dependency graph matrix representation. However, scipy sparse arrays are just sparse matrices with wrappers on top, so we can just get back to the matrix for now. While they claim they are deprecating the API, much of the API still works with and produces matrices, so I think we have time where we can continue doing what works.

In the long run, it would be nice to understand what is treating these representations differently, but that's beyond the time I have to devote to this. I would rather get things working before embarking on a refactor, especially if it's just a cheap type coercion.

Fixes #3537.
Fixes #3530.
Fixes #3539.

## List of changes proposed in this PR (pull-request)
* Convert `nx.to_scipy_sparse_matrix()` to `ssp.lil_matrix(nx.to_scipy_sparse_array())`
* Add non-int64 dtypes to arrays to be promoted to test Nifti1Images
* Fix dipy type conversion to accept `"str"` instead of `"string"`
* In passing update the requirements.txt file from current build info

### CODE REVIEW

**Comment by effigies** [comment]
Looks like dipy compatibility as well. @skoudoro anything we should know before digging in? I might have a chance this afternoon.

**Comment by skoudoro** [comment]
Nothing particular change. Looking at the error, it seems that a new workflow has a unknown argument type which surprise me. 

I can look at this only tonight, quite busy today.

Thank you for pinging @effigies

**Comment by effigies** [comment]
Thanks. I'll let you know if I find anything. If it's just a bug, the easy fix here will be to blacklist the latest release of dipy. Then there shouldn't be a huge rush for you to release.

**Comment by skoudoro** [comment]
Ok, I just looked deeper and you just need to add the type "str" in the function "convert_to_traits_type". This function is handling the keyword "string" but not "str". 😅

It should be an easy and fast fix of this function in nipype. Now, I see why I didn't catch it in DIPY.

File: nipype/interfaces/dipy/base.py [hunk 1]
```diff
@@ -110,7 +110,7 @@ def convert_to_traits_type(dipy_type, is_file=False):
     """Convert DIPY type to Traits type."""
     dipy_type = dipy_type.lower()
     is_mandatory = bool("optional" not in dipy_type)
-    if "variable" in dipy_type and "string" in dipy_type:
+    if "variable" in dipy_type and "str" in dipy_type:
```
Comment (effigies) [inline nipype/interfaces/dipy/base.py:113]: I tested this on 1.5.0 (which was not failing) and 1.6.0 (which was). Both pass.

**Review by TheChymera — APPROVED** [review]
LGTM, thanks for the amazing work. Solves the issue I just opened and supersedes my networkx PR.

(I still have [one error](https://github.com/nipy/nipype/issues/3540), but that is fully unrelated to dependency compatibility fixes)

**Pull request merged**